### ISSUE TITLE
Expected dev flow

### ISSUE THREAD

**Issue by farzadmf** [pulumi/pulumi#4772]
Hi,

DISCLAIMER: I just started playing with `pulumi` a couple of days ago, so sorry if my questions (and discussion) below are stupid. Please let me know if I'm missing something.

---

As a user of Terraform, I'm trying to figure out the expected flow for Pulumi, compared to Terraform.

Although I'm mainly focused on using an Azure backend (not `app.pulumi`), but I don't think that's relevant, and it doesn't change the concept I'm discussing below.

One thing for Pulumi that's a bit confusing for me is that you need access to all the backend information (to store files and encrypt secrets) at, what I call, "compile time" (ie, when creating the project and stacks).

One good thing I find with Terraform is that, at "compile time", you don't need to be connected to anything and you just say what you want to use, and it's only at "run time" where you connect to your backends and do stuff.

Let's say I want to create a project, with one `dev` and one `production` environments using Azure backends.

Question: If I'm not mistaken, both of them have to use the same blob for state storage (because they're in the same project), am I right?

That means that when creating the project, I need to run `pulumi login` to connect to my Azure backend, and I would need to be connected to my "production" backend (which, in my case, is not possible because only a handful of people have the information for that backend)

Can you please tell me what the intended flow is in these scenarios. The fact that you need to have access to everything at "compile time" for `pulumi` makes a bit harder to adopt (at least in my case for the production environments) compared to Terraform where you can provide all the information without being connected to anything, and it's only at "runtime" (when you want to _actually_ do something to do your environment) that you need to provide the information and be connected

**Comment by leezen** [comment]
You should think of the state store as separate from the actual providers you'll connect to. The state backend is simply where you store state and has nothing to do with where you'll actually deploy any resources managed by Pulumi.

For example, you could use GCS to store state and have your Pulumi program deploy your resources to Azure. Similarly, you could choose to store state in Azure with one set of credentials, but deploy your resources using a different set of credentials. If you'd prefer to completely manage the state yourself without any online interactions, you can always use the 'local' (filesystem) backend.

Hopefully that helps clarify, but please re-open if it does not.

**Comment by farzadmf** [comment]
Thank you @leezen for your answer,

I think my question was a bit misleading. Let me rephrase:

(Also, just to make discussions easier, let's consider that everything is in Azure: state storage and the actual deployment.)

I understand that the state store has nothing to do with my actual deployment. But let's say this is the scenario (matching what we currently have with Terraform):
- For "dev" environments, I want to use a "dev" storage, whose information (namely `AZURE_STORAGE_KEY`) is available (or I can do `az login` and use that)
- For "prod" environments, I want to use a "prod" storage, whose information (access key) is not available to me

Also, for secret management, I want to use a "dev" and "prod" Azure KeyVault with the same restrictions I explained above:
- In `pulumi`, you set `AZURE_KEYVAULT_AUTH_VIA_CLI` so that you use your `az` login information to access the KeyVault
- This is fine for the "dev" KeyVault, but for the "prod" KeyVault, I don't have access to the information required to log in to the Azure Subscription to have `az` login information to be used for KeyVault access

So, I was wondering how the scenario would work. Basically, in Terraform, you tell Terraform: "hey, get your Azure subscription and storage information from `x` and `y` etc. variables", and then you define your environment, put it into source code, and it's only upon the actual deployment (normally on a build machine) that you specify those variables and things happen. In other words, it's only when you want to do the actual `plan` and `apply` that you specify those `x`, `y`, etc. variables.

If I could use another analogy, I feel like with Terraform, you can have a "passive" connection to remote (that only becomes "active" when `plan`ing and `apply`ing), but in `pulumi` you need an "active" connection right from the start.

Sorry if my comments are a bit long; I'm trying to clearly explain what I mean

**Comment by leezen** [comment]
Thanks for the clarification and apologies for misunderstanding your initial post. I see what you mean by the idea around needing to be 'connected' when running a `pulumi stack init`

This basically comes down to a design decision around making state storage more integrated with the overall experience. In your example above, if you want to do some local development and then run it in your CI system, without a way to coordinate that state (e.g. `pulumi login`) you would have to have some way to encrypt those secrets and store them into your configuration (which `pulumi config` does for you). So in many ways, this is simply the design of making it easier to coordinate state, including secrets.

In terms of your scenario of having separate "dev" and "prod" setups -- there's nothing preventing you from doing so. Stacks for the same project don't need to be managed by the same backend. You can imagine a situation where locally, you develop against the `dev` stack and in your CI environment, it simply executes the same program (project) but against some other stack managed by some other backend. Or am I missing something in your scenario where this isn't possible?

**Comment by farzadmf** [comment]
> apologies for misunderstanding your initial post

No worries, I think my wording for the question was off too :slightly_smiling_face: 

I was thinking about this scenario:
Let's say I don't have any secrets as part of `pulumi config`, but I do want the outputs of my resources to be encrypted (in the state file).

Is it possible to do the creation of the stack(s) offline (`--local` and a simple passphrase), but when doing in a pipeline (CI/CD), I would log in to my remote storage and do `stck init --secrets-provider=azurekeyvault://...` and then do the actual deployment?

Oh but I'm wondering what happens in the case of changing/updating the deployment configuration :thinking: 

What I'm trying to achieve here is to be able to do things locally *without* doing any actual deployment; I'm only planning on changing the code for the deployment, adding/removing/updating resources, and then push to Git, and only then (in my CI/CD pipeline) take action and do the deployment.

And, again, the reason I'm trying to find these workarounds is that there's no way in a million years that I would have the information required to `az login` to a subscription that has access to "prod" KeyVault and Storage.

Any thoughts on this?

**Comment by leezen** [comment]
A couple thoughts:
1. If you want to store state in something only accessible via keys you can't access, then by definition, there's no way for you to interact with the resulting state locally. However, I assume you're okay not being able to interact with the state file (given your description of the setup above). In this case, there's no problem developing your program locally, editing config locally, and have that all "work" (obviously previews won't work since you have no access to state). Of course, this means you won't be able to create secrets in the config since you have no access to the prod KeyVault. But otherwise, plaintext config should work.
2. If you needed to manage secrets locally and have them accessible in prod, then by definition, you need to have a way to encrypt/decrypt the secrets the same way prod would do so. Since you don't have access to the prod KeyVault, that's not really an option. One alternative would be to use the passphrase secrets provider.

**Comment by farzadmf** [comment]
Thank you @leezen for your thoughts.

Let me go back to my original question and the reason I have "intended" in my issue header.

First of all, let me mention that I don't find what they do in my company (no trust on devs) "normal" AT ALL; it's just something I have to deal with, unfortunately.

I know `pulumi` is mainly designed to work with the  `app.pulumi` backend; this is something you feel when reading the docs (and I think I read somewhere that it's only been recently that it supports other backends)

What I'm trying to understand is the ideology and thought process of the flow that we should have when creating environment deployments with `pulumi`.

First thing in `pulumi` is to `pulumi login` to a backend. I'm not sure how `app.pulumi` works, but let's say for Azure, you would need an account access key, which would give you permission to change what's already there in the storage. Is the mindset in `pulumi` is "we trust the person creating the deployment not to do that"? (again, I emphasize that I'm not saying that this mindset is bad, and, in fact, it's personally my ideal mindset; I'm just trying to understand the concept)

And, then, let's say we want to configure our `stack` to use Azure KeyVault for encryption. Then, you would need to either `az login` with an account with access to the KeyVault, or have the required information to use for [Azure Environment Authentication](https://docs.microsoft.com/en-us/azure/developer/go/azure-sdk-authorization#use-environment-based-authentication). In both cases, that would mean that the person has access to some "sensitive" information. What's the mindset here? Do we again trust the person to have that information?

Sorry again for my long comments. When I became familiar with `pulumi` a few days ago, I instantly fell in love!! and I found it a tool that blows Terraform out of the water, and, honestly, I was planning to do experiments and then let people know of all the BIG advantages over Terraform.

And, the reason that I'm bothering you with my questions is that I think `pulumi` is too good for me to give up because of the crazy limitations there are in my company, and it makes me sad (literally!) if I can't use `pulumi` instead of Terraform.

**Comment by leezen** [comment]
It's not about trust/no trust -- you'll find the same issues in Terraform except that it won't be Terraform per se, but more likely Vault. If you don't have access to your 'prod' Vault, how are you going to modify the configuration for it?

If the answer to the previous question is "some admin is going to do it" you could use the same workflow in Pulumi. For example, you could have the admin create whatever secrets are needed and when CI executes, it will select the "prod" stack and use the "prod" secrets (configured by your admin). When you are developing locally against your "dev" stack, you'll have your own set of secrets and config. When you check in your code, it's the same "program" that's executed against both stacks. Does that hopefully make sense?

**Comment by farzadmf** [comment]
@leezen , I'm trying my best not to make a fool of myself by asking the wrong questions.

But I think (and I hope) I see where you're going. Let me try to re-iterate and please correct me if I'm wrong.

If I'm not mistaken, this would be a flow (again, assuming we want to use Azure as backend and Azure KeyVault for encryption):
- We have our repository containing code and deployment
- A person with access to the KeyVault creates a stack (eg. `qa`) with this: `pulumi stack init qa --secrets-provider=azurekeyvault://...`
- This creates the `Pulumi.qa.yaml` in your repository (which I think by default only contains `secretsprovider` and `encryptedkey` fields)
- They commit and push this file into `git`
- The person without KeyVault access pulls this file and can start adding configuration to the file

Since they're not going to do any deployment locally (ie, running any `pulumi` commands), they should be fine without access to the KeyVault. Although I think if they want to do `pulumi config set ... --secret`, they would need access to the KeyVault and again, that should be done by a person with access to they KeyVault.

During CI, you just need to make sure that the pipeline has access to they KeyVault. You `pulumi login` to your backend, `pulumi stack select` the proper stack and trigger the deployment

Did I understand what you mean? Am I on the right track?

**Comment by leezen** [comment]
Yes -- this is more or less what I had in mind in the scenario where you (the developer without access to `qa` access) would still be able to interact with the configuration.

**Comment by farzadmf** [comment]
Great @leezen :grinning: that's very good to hear, I think that's something I can work with and maneuver around. I'm going to do it, and hopefully start using `pulumi` very soon (I'm going to bother you hear in case there are blockers :stuck_out_tongue_closed_eyes: )

Also, just out of curiosity, I tried to find it on the internet, but I wasn't able to; I was trying to find the origin(s) for the word `pulumi` (what it means, why chosen as the product name), but I didn't find anything

**Comment by leezen** [comment]
Glad to hear. You'll find plenty of helpful folks on our Community Slack (slack.pulumi.com)

As for the origin of the name, Joe talks about in the PS of this blog post: http://joeduffyblog.com/2018/06/18/hello-pulumi/

**Comment by farzadmf** [comment]
@leezen First of all, thank you for the link; it was very cool to read (and it has an emotional sense to it when you know where Pulumi comes from ❤️ )

But, I'm back with my questions 🙂 , I didn't open a new issue because I feel it could be categorized as the same "intended flow" that I was asking.

Again, please note that the things I'm saying are in the context of using an Azure Storage backend

I'm a bit unsure of how to properly use stacks. Let's say we have 2 different services and they both want to use the same backend. My thinking about stacks was that they would somehow represent your different environments for your service (`dev`, `qa`, `prod`, etc.), but that would mean that for 2 different services, you need to make sure they use different Azure Blobs because if you want to do `pulumi stack init` in one, it would complain that the stack already exists.

Or this is the expected configuration: each service needs to use its own blob, and they shouldn't share their backend?

I feel like I'm missing something here, and my thought process is not right. I would appreciate if you guide me here 🙂

**Comment by leezen** [comment]
If you haven't already, I'd suggest reading https://www.pulumi.com/docs/intro/concepts/organizing-stacks-projects/
(and make any suggestions for how we can improve that page!)

In general, I tend of think of a "project" as something that encapsulates a concept or single "unit" of infrastructure and the "stacks" of that project as individual instantiations of that concept.

One way to achieve your example of two different services sharing a backend, would be to have a "backend" project along with a project for service `A` and a project for service `B`. You could have `A` and `B` both have stack references to the backend stack(s). For example, you could imagine something like (I'll use `project/stack` annotation below):

`A/prod` stack references `backend/shared`
`A/dev` stack references `backend/shared`
`B/prod` stack references `backend/shared`
`B/dev` stack references `backend/shared`

You could also imagine another configuration where
`A/prod` stack references `backend/prod`
`A/dev` stack references `backend/dev`
`B/prod` stack references `backend/prod`
`B/dev` stack references `backend/dev`

**Comment by farzadmf** [comment]
Thank you @leezen for your answer. I had the link before, and I'm still not sure about the concepts.

Would you please elaborate a bit about what you mean when you say "`A` and `B` both have stack references to the backend stack(s)"?

Also, I _think_ (I'm not sure) that most of these concepts make sense in the context of using `app.pulumi` backend. For example, I think the concept of a "project" _only_ makes sense in that context because if you use, for example, `login --local`, all you have is some directories in `~/.pulumi`, where I think the only interesting ones are `stacks` and `workspaces` (I'm not exactly sure what `workspaces` is for)

I haven't tried with other backends, but at least for Azure, I've seen the same folder structure as `local`. For this reason, I don't think the concept of a project really applies to other backends, and the only thing you have there to play with are the stacks.

So, I'm a bit confused here, and I'm not sure what's the recommended practice is when you don't use the `app.pulumi` backend

**Comment by leezen** [comment]
You can still think of projects and stacks in the same way -- you just can't name them quite the same way as I'm describing.

**Comment by farzadmf** [comment]
Oh OK, I think I see what you mean, so if I'm not mistaken, you're suggesting that: let's say, we create `A` using a fresh Azure Storage Container and a `dev` stack (for example), then, we go to service `B` and login to the same backend (Azure Container), and then we reference the `dev` stack there. Is that what you meant?

What I tried just now was trying to do `pulumi login` to my Azure Storage, then creating a new "project" by doing `pulumi new ... --stack=dev` (the stack my other project is using), but it gives me an error indicating the stack already exists. This is also a bit confusing for me given that in the help it says:
```
  -s, --stack string              The stack name; either an existing stack or stack to create; if not specified, a prompt will request it
```
where it says an _existing_ stack or create

**Comment by leezen** [comment]
So instead of naming your stack `dev` you would probably adopt a naming scheme like `<project>-<stack>` and then both `A-dev` and `B-dev` will reference `backend-dev` for example

**Comment by farzadmf** [comment]
Yup, I joined 10 seconds after your comment :smile:
